odoo/addons/test_new_api/tests/test_company_checks.py

```diff
@@ -72,3 +72,51 @@ def test_company_check_4(self):
             'company_id': self.company_b.id,
         })
 
+    def test_company_environment(self):
+        """ Check the company context on the environment is verified. """
+
+        self.company_c = self.env['res.company'].create({
+            'name': 'Company C'
+        })
+
+        user = self.env['res.users'].create({
+            'name': 'Test',
+            'login': 'test',
+            'company_id': self.company_a.id,
+            'company_ids': (self.company_a | self.company_c).ids,
+        })
+
+        user = user.with_user(user).with_context(allowed_company_ids=[])
+
+        # When accessing company/companies, check raises error if unauthorized/unexisting company.
+        with self.assertRaises(AccessError):
+            user.with_context(allowed_company_ids=[self.company_a.id, self.company_b.id, self.company_c.id]).env.companies
+
+        with self.assertRaises(AccessError):
+            user.with_context(allowed_company_ids=[self.company_b.id]).env.company
+
+        with self.assertRaises(AccessError):
+            # crap in company context is not allowed.
+            user.with_context(allowed_company_ids=['company_qsdf', 'company564654']).env.companies
+
+        # In sudo mode, context check is bypassed.
+        companies = (self.company_a | self.company_b)
+        self.assertEqual(
+            user.sudo().with_context(allowed_company_ids=companies.ids).env.companies,
+            companies
+        )
+
+        self.assertEqual(
+            user.sudo().with_context(
+                allowed_company_ids=[self.company_b.id, 'abc']).env.company,
+            self.company_b
+        )
+        """
+        wrong_env = user.sudo().with_context(
```

Comment:
remove?

Reply:
It is a highlight to clearly say that it is intended to be so, even if not testable.
If you think it doesn't help, I can remove it.